pki/leaf.go

```diff
@@ -0,0 +1,271 @@
+
+	RemoveDNSNames(...string)
+
+	RemoveIPAddresses(...net.IP)
+
+	Signer() crypto.Signer
+}
+
+type ipLease struct {
+	Address  net.IP
+	RefCount uint
+}
+
+func (d *DefaultLeaf) addDNSNames(register bool, dnsNames ...string) {
+	d.changeLock.Lock()
+	defer d.changeLock.Unlock()
+	for _, dnsName := range dnsNames {
+		if v, exists := d.dnsNames[dnsName]; exists {
+			if register {
+				d.dnsNames[dnsName] = v + 1
+			}
+			continue
+		}
+		d.dnsNames[dnsName] = 1
+	}
+}
+
+func (d *DefaultLeaf) addIPAddresses(register bool, addresses ...net.IP) {
+	d.changeLock.Lock()
+	defer d.changeLock.Unlock()
+	for _, address := range addresses {
+		if v, exists := d.ipAddresses[address.String()]; exists {
+			if register {
+				v.RefCount = v.RefCount + 1
+			}
+			continue
+		}
+		d.ipAddresses[address.String()] = &ipLease{
+			Address:  address,
+			RefCount: 1,
+		}
+	}
+}
+
+func (d *DefaultLeaf) AddDNSNames(dnsNames ...string) error {
+	d.addDNSNames(false, dnsNames...)
+	return nil
+}
+
+func (d *DefaultLeaf) AddIPAddresses(addresses ...net.IP) error {
+	d.addIPAddresses(false, addresses...)
+	return nil
+}
+
+func (d *DefaultLeaf) Certificate() *x509.Certificate {
+	d.certificateLock.RLock()
+	defer d.certificateLock.RUnlock()
+	return d.certificate
+}
+
+func (d *DefaultLeaf) Chain() []*x509.Certificate {
+	d.certificateLock.RLock()
+	defer d.certificateLock.RUnlock()
+	return d.certificateChain
+}
+
+func (d *DefaultLeaf) Commit() error {
+	d.certificateLock.Lock()
+	defer d.certificateLock.Unlock()
+	csr := &x509.CertificateRequest{
+		DNSNames:    dnsMapToSlice(d.dnsNames),
+		IPAddresses: ipAddressMapToSlice(d.ipAddresses),
+		PublicKey:   d.signer.Public(),
+		Subject:     d.subject,
+	}
+	cert, chain, err := d.requestSigner.SignCSR(csr)
+	if err != nil {
+		return errors.Annotatef(err, "signing CSR for leaf group %s", d.Group)
+	}
+	d.certificate = cert
+	d.certificateChain = chain
+	return nil
+}
+
+func dnsMapToSlice(m map[string]uint) []string {
+	rval := make([]string, len(m))
+	i := 0
+	for k, _ := range m {
+		rval[i] = k
+		i = i + 1
+	}
+	return rval
+}
+
+func (d *DefaultLeaf) DNSNames() []string {
+	d.changeLock.Lock()
+	defer d.changeLock.Unlock()
+	return dnsMapToSlice(d.dnsNames)
+}
+
+func ipAddressMapToSlice(m map[string]*ipLease) []net.IP {
+	rval := make([]net.IP, len(m))
+	i := 0
+	for _, v := range m {
+		rval[i] = v.Address
+		i = i + 1
+	}
+	return rval
+}
+
+func (d *DefaultLeaf) IPAddresses() []net.IP {
+	d.changeLock.Lock()
+	defer d.changeLock.Unlock()
+	return ipAddressMapToSlice(d.ipAddresses)
+}
+
+func NewDefaultLeaf(group string, subject pkix.Name,
+	requestSigner CertificateRequestSigner) (*DefaultLeaf, error) {
+
+	signer, err := DefaultKeyProfile()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &DefaultLeaf{
+		certificateLock: sync.RWMutex{},
+		changeLock:      sync.Mutex{},
+		dnsNames:        map[string]uint{},
+		Group:           group,
```

Comment:
not sure why Group is exported

Reply:
The plan was to use it in pem tags. But that never happened